README.md
# influxdb-relay

[![License][license-img]][license-href]

1. [Overview](#overview)
2. [Description](#description)
3. [Requirements](#requirements)
4. [Setup](#setup)
5. [Usage](#usage)
6. [Limitations](#limitations)
7. [Development](#development)
8. [Miscellaneous](#miscellaneous)

## Overview

Maintained fork of [influxdb-relay][overview-href] originally developed by
InfluxData. Replicate InfluxDB data for high availability.

**Be careful, we have deeply changed the configuration file syntax.**

## Description

This project adds a basic high availability layer to InfluxDB. With the right
architecture and disaster recovery processes, this achieves a highly available
setup.

## Tested on

- [Go](https://golang.org/doc/install) 1.7.4 to 1.12
- [InfluxDB](https://docs.influxdata.com/influxdb/v1.7/introduction/installation/) 1.5 to 1.7 (2.x not yet supported)

Other versions will probably work but are untested.

## Setup

### Go

Download the daemon into your `$GOPATH` and install it in `/usr/sbin`.

```sh
go get -u github.com/vente-privee/influxdb-relay
cp ${GOPATH}/bin/influxdb-relay /usr/bin/influxdb-relay
chmod 755 /usr/bin/influxdb-relay
```

Create the configuration file in `/etc/influxdb-relay`.

```sh
mkdir -p /etc/influxdb-relay
cp ${GOPATH}/src/github.com/vente-privee/influxdb-relay/examples/sample.conf \
   /etc/influxdb-relay/influxdb-relay.conf
```

### Docker

Build your own image.

```sh
git clone git@github.com:vente-privee/influxdb-relay
cd influxdb-relay
docker build \
       --file Dockerfile \
       --rm \
       --tag influxdb-relay:latest \
       .
docker run \
       --volume /path/to/influxdb-relay.conf:/etc/influxdb-relay/influxdb-relay.conf \
       --publish 9096:9096 \
       --rm \
       influxdb-relay:latest
```

or

Docker pull our image.

```sh
docker pull vptech/influxdb-relay:latest
docker run \
       --volume /path/to/influxdb-relay.conf:/etc/influxdb-relay/influxdb-relay.conf \
       --publish 9096:9096 \
       --rm \
       vptech/influxdb-relay:latest
```

## Usage

You can find more documentation in [docs](docs) folder.

* [Architecture](docs/architecture.md)
* [Buffering](docs/buffering.md)
* [Caveats](docs/caveats.md)
* [Recovery](docs/recovery.md)
* [Filters](docs/filters.md)
* [Sharding](docs/sharding.md)

You can find some configurations in [examples](examples) folder.

* [sample.conf](examples/sample.conf)
* [sample_buffered.conf](examples/sample_buffered.conf)
* [kapacitor.conf](examples/kapacitor.conf)

### Configuration

```toml
[[http]]
# Name of the HTTP server, used for display purposes only.
name = "example-http"

# TCP address to bind to, for HTTP server.
bind-addr = "0.0.0.0:9096"

# Timeout for /health route
# After this time, the host may be considered down
health-timeout-ms = 10000

# Request limiting (Applied to all backend)
rate-limit = 5
burst-limit = 10

# Ping response code, default is 204
default-ping-response = 200

# Enable HTTPS requests.
ssl-combined-pem = "/path/to/influxdb-relay.pem"

# InfluxDB instances to use as backend for Relay
[[http.output]]
# name: name of the backend, used for display purposes only.
name = "local-influxdb01"

# location: full URL of the /write endpoint of the backend
location = "http://127.0.0.1:8086/"

# endpoints: Routes to use on Relay
# write: Route for standard InfluxDB request
# write_prom: Route for Prometheus request
# ping: Route for ping request
# query: Route fot querying InfluxDB backends
endpoints = {write="/write", write_prom="/api/v1/prom/write", ping="/ping", query="/query"}

# timeout: Go-parseable time duration. Fail writes if incomplete in this time.
timeout = "10s"

# skip-tls-verification: skip verification for HTTPS location. WARNING: it's insecure. Don't use in production.
skip-tls-verification = false

# InfluxDB
[[http.output]]
name = "local-influxdb02"
location = "http://127.0.0.1:7086/"
endpoints = {write="/write", ping="/ping", query="/query"}
timeout = "10s"

# Prometheus
[[http.output]]
name = "local-influxdb03"
location = "http://127.0.0.1:6086/"
endpoints = {write="/write", write_prom="/api/v1/prom/write", ping="/ping", query="/query"}
timeout = "10s"

[[http.output]]
name = "local-influxdb04"
location = "http://127.0.0.1:5086/"
endpoints = {write="/write", write_prom="/api/v1/prom/write", ping="/ping", query="/query"}
timeout = "10s"

[[udp]]
# Name of the UDP server, used for display purposes only.
name = "example-udp"

# UDP address to bind to.
bind-addr = "0.0.0.0:9096"

# Socket buffer size for incoming connections.
read-buffer = 0 # default

# Precision to use for timestamps
precision = "n" # Can be n, u, ms, s, m, h

# InfluxDB instance to use as backend for Relay.
[[udp.output]]
# name: name of the backend, used for display purposes only.
name = "local-influxdb01"

# location: host and port of backend.
location = "127.0.0.1:8089"

# mtu: maximum output payload size
mtu = 512

[[udp.output]]
name = "local-influxdb02"
location = "127.0.0.1:7089"
mtu = 1024
```

InfluxDB Relay is able to forward from a variety of input sources, including:

* `influxdb`
* `prometheus`

### Administrative tasks

#### /admin endpoint

Whereas data manipulation relies on the `/write` endpoint, some other features
such as database or user management are based on the `/query` endpoint. As
InfluxDB Relay does not send back a response body to the client(s), we are not
able to forward all of the features this endpoint provides. Still, we decided
to expose it through the `/admin` route.

It is now possible to query the `/admin` endpoint. Its usage is the same as the
standard `/query` Influx DB enpoint:

```
curl -X POST "http://127.0.0.1:9096/admin" --data-urlencode 'q=CREATE DATABASE some_database'
```

Errors will be logged just like regular `/write` queries. The HTTP response
bodies will not be forwarded back to the clients.

#### /health endpoint

This endpoint provides a quick way to check the state of all the backends.
It will return a JSON object detailing the status of the backends like this :

```json
{
  "status": "problem",
  "problem": {
    "local-influxdb01": "KO. Get http://influxdb/ping: dial tcp: lookup influxdb on 0.0.0.0:8086: no such host"
  },
  "healthy": {
    "local-influxdb02": "OK. Time taken 3ms"
  }
}
```

If the relay encounters an error while checking a backend, this backend will be reported with the associated error in the `problems` object.
The backends wich the relay was able to communicate with will be reported in the healthy object.

The status field is a summary of the general state of the backends, the defined states are as follows:
* `healthy`: no errors were encountered
* `problem`: some backends, but no all of them, returned errors
* `critical`: every backend returned an error

### Filters

We allow tags and measurements filtering through regular expressions. Please,
take a look at [this document](docs/filters.md) for more information.

## Limitations

So far, this is compatible with Debian, RedHat, and other derivatives.

## Development

Please read carefully [CONTRIBUTING.md][contribute-href] before making a merge
request.

Clone repository into your `$GOPATH`.

```sh
mkdir -p ${GOPATH}/src/github.com/vente-privee
cd ${GOPATH}/src/github.com/vente-privee
git clone git@github.com:vente-privee/influxdb-relay
```

Enter the directory and build the daemon.

```sh
cd ${GOPATH}/src/github.com/vente-privee/influxdb-relay
go build -a -ldflags '-extldflags "-static"' -o influxdb-relay
```

## Miscellaneous

```
    ╚⊙ ⊙╝
  ╚═(███)═╝
 ╚═(███)═╝
╚═(███)═╝
 ╚═(███)═╝
  ╚═(███)═╝
   ╚═(███)═╝
```

[license-img]: https://img.shields.io/badge/license-MIT-blue.svg
[license-href]: LICENSE
[overview-href]: https://github.com/influxdata/influxdb-relay
[contribute-href]: CONTRIBUTING.md


CHANGELOG.md
Dec 6 2019 Damien PLENARD <dplenard@vente-privee.com>
	* Merge https://github.com/strike-team/influxdb-relay/pull/38
    * Merge https://github.com/strike-team/influxdb-relay/pull/36

Jun 28 2018 Antoine MILLET <amillet@vente-privee.com>
	From https://github.com/influxdata/influxdb-relay

Jun 29 2018 Alexandre BESLIC <abeslic@abronan.com>
	* Switch to dep, make the project buildable after go get

Jun 29 2018 Dejan FILIPOVIC <dfilipovic@vente-privee.com>
	* Add /status route

Jun 30 2018 Antoine MILLET <amillet@vente-privee.com>
	* New README structure
	* Add basic tests with golint & pylint
	* Add CHANGELOG
	* Add CONTRIBUTING guide
	* Merge https://github.com/influxdata/influxdb-relay/pull/65
	* Merge https://github.com/influxdata/influxdb-relay/pull/52
	* Merge https://github.com/influxdata/influxdb-relay/pull/59
	* Merge https://github.com/influxdata/influxdb-relay/pull/43
	* Merge https://github.com/influxdata/influxdb-relay/pull/57

Nov 15 2018 Maxime CORBIN <mcorbin@vente-privee.com>
    * Add Prometheus input support
    * Add `/admin` route to administrate underlying databases
    * Add code coverage / unit tests
    * Code refactor
    * Improve buffering feature avoiding connexions hanging
    * Improve `/ping` route
    * Improve logging
    * Add `-version` option

Dec 13 2018 Maxime CORBIN <mcorbin@vente-privee.com>
    * Add `/admin` endpoint that can be used to create or remove databases
    * Add `/health` endpoint that can be used to monitor the health status of every backend
    * Fixed some performance bugs & added a few more logs

Dec 27 2018 Clément CORNUT <ccornut@vente-privee.com>
    * Add `/admin/flush` endpoint that can be used to flush internal buffer
    * Add Rate limiting on backend
    * Add Filters for tags and measurements
    * New configuration formatting
    * Various bug fixes

config/config.go
package config

import (
	"log"
	"os"
	"regexp"

	"github.com/naoina/toml"
)

// Config is an object created from a configuration file
// It is a list of HTTP and/or UDP relays
// Each relay has its own list of backends
type Config struct {
	HTTPRelays []HTTPConfig `toml:"http"`
	UDPRelays  []UDPConfig  `toml:"udp"`
	Filters    Filters      `toml:"filter"`
	Verbose    bool
}

// Filter represents a regex which may be
// applied to the incoming requests
type Filter struct {
	// Type is how the regex result will be interpreted
	Type string `toml:"type"`

	// TagExpression is a valid Go regex
	// It will be applied on each tag of any request to the backend
	TagExpression string `toml:"tag-expression"`

	// MeasurementExpression is a valid Go regex
	// It will be applied on the measurement of any request to the backend
	MeasurementExpression string `toml:"measurement-expression"`

	// TagRegexp is the compiled tag regexp
	TagRegexp *regexp.Regexp

	// MeasurementRegexp is the compiled measurement regexp
	MeasurementRegexp *regexp.Regexp

	// Outputs are the endoints the regex are applied on
	Outputs []string `toml:"outputs"`
}

// Filters is a type representing an array of Filter, wow
type Filters []Filter

// HTTPConfig represents an HTTP relay
type HTTPConfig struct {
	// Name identifies the HTTP relay
	Name string `toml:"name"`

	// Addr should be set to the desired listening host:port
	Addr string `toml:"bind-addr"`

	// Set certificate in order to handle HTTPS requests
	SSLCombinedPem string `toml:"ssl-combined-pem"`

	// Default retention policy to set for forwarded requests
	DefaultRetentionPolicy string `toml:"default-retention-policy"`

	DefaultPingResponse int `toml:"default-ping-response"`

	// Rate limit on specific HTTP relay
	RateLimit int `toml:"rate-limit"`

	// Burst allowed by rate limit
	BurstLimit int `toml:"burst-limit"`

	// Outputs is a list of backed servers where writes will be forwarded
	Outputs []HTTPOutputConfig `toml:"output"`

	HealthTimeout int64 `toml:"health-timeout-ms"`
}

// HTTPOutputConfig represents the specification of an HTTP backend target
type HTTPOutputConfig struct {
	// Name of the backend server
	Name string `toml:"name"`

	// Location should be set to the hostname for the influxdb endpoint (for example https://influxdb.com/)
	Location string `toml:"location"`

	// Endpoints should contain the path to the different influxdb endpoints used
	Endpoints HTTPEndpointConfig `toml:"endpoints"`

	// Location should be set to the hostname for the influxdb endpoint (for example https://influxdb.com/)
	Urltype string `toml:"url-type"`

	// Timeout sets a per-backend timeout for write requests (default: 10s)
	// The format used is the same seen in time.ParseDuration
	Timeout string `toml:"timeout"`

	// Buffer failed writes up to maximum count (default: 0, retry/buffering disabled)
	BufferSizeMB int `toml:"buffer-size-mb"`

	// Maximum batch size in KB (default: 512)
	MaxBatchKB int `toml:"max-batch-kb"`

	// Maximum delay between retry attempts
	// The format used is the same seen in time.ParseDuration (default: 10s)
	MaxDelayInterval string `toml:"max-delay-interval"`

	// Skip TLS verification in order to use self signed certificate
	// WARNING: It's insecure, use it only for developing and don't use in production
	SkipTLSVerification bool `toml:"skip-tls-verification"`
}

//HTTPEndpointConfig details the remote endpoints to use
type HTTPEndpointConfig struct {
	// Must be the standard write endpoint in influxdb.
	Write string `toml:"write"`
	// Must be the prometheus specific influxdb endpoint
	PromWrite string `toml:"write_prom"`
	// Must be the ping endpoint
	Ping string `toml:"ping"`
	// Must be the query influxdb endpoint
	Query string `toml:"query"`
}

// UDPConfig represents a UDP relay
type UDPConfig struct {
	// Name identifies the UDP relay
	Name string `toml:"name"`

	// Addr is where the UDP relay will listen for packets
	Addr string `toml:"bind-addr"`

	// Precision sets the precision of the timestamps (input and output)
	Precision string `toml:"precision"`

	// ReadBuffer sets the socket buffer for incoming connections
	ReadBuffer int `toml:"read-buffer"`

	// Outputs is a list of backend servers where writes will be forwarded
	Outputs []UDPOutputConfig `toml:"output"`
}

// UDPOutputConfig represents the specification of a UDP backend target
type UDPOutputConfig struct {
	// Name identifies the UDP backend
	Name string `toml:"name"`

	// Location should be set to the host:port of the backend server
	Location string `toml:"location"`

	// MTU sets the maximum output payload size, default is 1024
	MTU int `toml:"mtu"`
}

// LoadRegexps will try to compile all the eventual regular expressions
// for each filter
// Any error here is critical
func (fs Filters) LoadRegexps() error {
	var err error

	for i := range fs {
		f := &fs[i]
		f.TagRegexp, err = regexp.Compile(f.TagExpression)
		if err != nil {
			return err
		}

		f.MeasurementRegexp, err = regexp.Compile(f.MeasurementExpression)
		if err != nil {
			return err
		}
	}

	return nil
}

func checkDoubleSlash(endpoint HTTPEndpointConfig) HTTPEndpointConfig {
	if endpoint.PromWrite != "" && endpoint.PromWrite[0] == '/' {
		endpoint.PromWrite = endpoint.PromWrite[1:]
	}

	if endpoint.Ping != "" && endpoint.Ping[0] == '/' {
		endpoint.Ping = endpoint.Ping[1:]
	}

	if endpoint.Query != "" && endpoint.Query[0] == '/' {
		endpoint.Query = endpoint.Query[1:]
	}

	if endpoint.Write != "" && endpoint.Write[0] == '/' {
		endpoint.Write = endpoint.Write[1:]
	}

	return endpoint
}

// LoadConfigFile parses the specified file into a Config object
func LoadConfigFile(filename string) (Config, error) {
	var cfg Config

	f, err := os.Open(filename)
	if err != nil {
		return cfg, err
	}
	defer f.Close()

	err = toml.NewDecoder(f).Decode(&cfg)
	if err == nil {
		for i, r := range cfg.HTTPRelays {
			for j, b := range r.Outputs {
				log.Printf(" name : , %v , url : %v , type :  %v \n", b.Name, b.Location, b.Urltype)
				if b.Location[len(b.Location)-1] == '/' {
					cfg.HTTPRelays[i].Outputs[j].Endpoints = checkDoubleSlash(b.Endpoints)
				}
			}
		}
		err = cfg.Filters.LoadRegexps()
	}
	return cfg, err
}
